File: persistence/src/main/java/io/sparkled/persistence/song/impl/query/SaveSongQuery.java
package io.sparkled.persistence.song.impl.query;

import io.sparkled.model.entity.Song;
import io.sparkled.model.validator.SongValidator;
import io.sparkled.persistence.PersistenceQuery;

import javax.persistence.EntityManager;

public class SaveSongQuery implements PersistenceQuery<Integer> {

    private final Song song;

    public SaveSongQuery(Song song) {
        this.song = song;
    }

    @Override
    public Integer perform(EntityManager entityManager) {
        new SongValidator(song).validate();
        Song result = entityManager.merge(song);
        return result.getId();
    }
}

File: persistence/src/main/kotlin/io/sparkled/persistence/song/impl/query/SaveSongQuery.kt
package io.sparkled.persistence.song.impl.query

import io.sparkled.model.entity.Song
import io.sparkled.model.validator.SongValidator
import io.sparkled.persistence.PersistenceQuery
import io.sparkled.persistence.QueryFactory
import org.slf4j.LoggerFactory

class SaveSongQuery(private val song: Song) : PersistenceQuery<Song> {

    override fun perform(queryFactory: QueryFactory): Song {
        SongValidator().validate(song)

        val entityManager = queryFactory.entityManager
        val savedSong = entityManager.merge(song)

        logger.info("Saved song {} ({}).", savedSong.getId(), savedSong.getName())
        return savedSong
    }

    companion object {
        private val logger = LoggerFactory.getLogger(SaveSongQuery::class.java)
    }
}
